### PR TITLE
Fix use of default worker pool on workers.

### PR DESCRIPTION
**Pull request by amitmurthy**
Addresses https://github.com/JuliaLang/julia/issues/19262 for 0.6

0.5 will need a different fix as this patch will break compatibility between 0.5.0 and 0.5.x

### CODE REVIEW

File: base/workerpool.jl [hunk 1]
```diff
@@ -40,12 +43,18 @@ function WorkerPool(workers::Vector{Int})
...
 push!(pool::AbstractWorkerPool, w::Worker) = push!(pool, w.id)
+push!(pool::AbstractWorkerPool, w::Int) = (push!(pool.workers, w); put!(pool.channel, w); pool)
 length(pool::AbstractWorkerPool) = length(pool.workers)
 isready(pool::AbstractWorkerPool) = isready(pool.channel)
 
-put!(pool::AbstractWorkerPool, w::Int) = (put!(pool.channel, w); pool)
+function put!(pool::AbstractWorkerPool, w::Int)
+    # In case of default_worker_pool, the master is implictly considered a worker, i.e.,
+    # it is not present in pool.workers.
+    # Confirm the that tha worker is part of a pool before making it available.
```
Comment (tkelman) [inline base/workerpool.jl]: Confirm that the worker

**Comment by kshyatt** [comment]
Is anything holding this up?

**Comment by amitmurthy** [comment]
Updated the patch to support serializing custom worker pools across workers. Will keep it open for a couple of days and merge if CI is green.

**Comment by ViralBShah** [comment]
bump.

File: base/workerpool.jl [hunk 2]
```diff
@@ -92,14 +110,51 @@ function remotecall_pool(rc_f, f, pool::AbstractWorkerPool, args...; kwargs...)
...
 end
 
+# Check if pool is local or remote and forward calls if required.
+# NOTE: remotecall_fetch does it automatically, but this will be more efficient as
+# it avoids the overhead associated with a local remotecall.
+
+for func = (:length, :isready, :workers, :nworkers, :take!)
+    func_local = Symbol(string("wp_local_", func))
+    eval(
+        quote
```
Comment (tkelman) [inline base/workerpool.jl]: `@eval begin` would match the more commonly used style, wouldn't it?

**Comment by tkelman** [comment]
does this need a genstdlib run or are these not exported?

**Comment by amitmurthy** [comment]
Changed eval -> `@eval`

The functions are exported but not specifically documented in the context of Worker Pools. Using `WorkerPool` in the typical case (via `pmap`, `remote`, etc) does not require those functions. Should be done in a different PR as part of a more detailed documentation on lower level use of worker pools and writing custom worker pools. 

Previous CI errors were `spawn` failures. Can be merged once CI is green.

File: base/workerpool.jl [hunk 3]
```diff
@@ -11,59 +11,77 @@ abstract AbstractWorkerPool
...
+        wp = WorkerPool(Channel{Int}(typemax(Int)), RemoteChannel())
+        put!(wp.ref, wp)
+        wp
+    end
 
-Create a WorkerPool from a vector of worker ids.
-"""
-function WorkerPool(workers::Vector{Int})
-    pool = WorkerPool()
+    """
```
Comment (tkelman) [inline base/workerpool.jl]: This needs a `@doc` I believe, see the warning from `make docs` - and there was another change that needed a doc update for "used by ``remote(f)`` and ``pmap`` (by default)"

**Pull request merged**